### PR TITLE
feat: Delete logs support

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
Adds support for deletion of logs. The feature is still experimental so it is required to be enabled explicitly by setting `purger.enable` config to `true`.

**NOTE**: This feature is only supported when using index stores other than `boltdb-shipper`. Support for it would be added in a separate PR.

This feature works the same as [delete series support in Cortex](https://github.com/cortexproject/cortex/blob/6db67a4efbbf62b1133fa037a95382a21f752bbf/docs/guides/deleting-series.md), the only change is the URLs used for performing various operations, which are outlined below:

* Requesting Deletion:
```
PUT,POST <loki-addr>/loki/api/admin/delete?match[]=<matcher>&start=<start-time>&end=<end-time>
```

* List Delete Requests:
```
PUT,POST <loki-addr>/loki/api/admin/delete
```

* Cancelling a Delete Requests:
```
PUT,POST <loki-addr>/loki/api/admin/delete?request_id=<request-id>
```
Request ID of the request can be grabbed from the List Delete Request API response.

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: pkg/iter/entry_iterator.go [hunk 1]
```diff
@@ -743,3 +747,120 @@ func (it *peekingEntryIterator) Error() error {
...
+				continue
+			} else if tm.After(interval.End) {
+				return false
+			}
+
+			return true
+		}
+	}
+
+	return false
```
Comment (cyriltovena) [inline pkg/iter/entry_iterator.go:830]: you should try to remove all the nesting here, by using another function.

File: pkg/loki/modules.go [hunk 2]
```diff
@@ -495,6 +499,60 @@ func (t *Loki) initCompactor() (services.Service, error) {
...
+	}
+
+	t.Purger, err = purger.NewPurger(t.cfg.PurgerConfig, t.DeletesStore, t.store, storageClient, prometheus.DefaultRegisterer)
+	if err != nil {
+		return nil, err
+	}
+
+	deleteRequestHandler := loki_purger.NewDeleteRequestHandler(t.DeletesStore, t.cfg.PurgerConfig.DeleteRequestCancelPeriod, prometheus.DefaultRegisterer)
+
+	t.server.HTTP.Path("/loki/api/admin/delete").Methods("PUT", "POST").Handler(t.httpAuthMiddleware.Wrap(http.HandlerFunc(deleteRequestHandler.AddDeleteRequestHandler)))
```
Comment (cyriltovena) [inline pkg/loki/modules.go:549]: I'm not sure about those path, I would have expected a DELETE method :) @slim-bean wdyt is that how cortex does it ?

Comment (cyriltovena) [reply pkg/loki/modules.go:549]: We should talk about this, it's very easy to change anyway, but it's better now than later.

Comment (sandeepsukhani) [reply pkg/loki/modules.go:549]: This is how Cortex does which is inspired from Prometheus. We can add support for DELETE method as well.

Comment (cyriltovena) [reply pkg/loki/modules.go:549]: Can you link me to Prometheus and Cortex docs please ?

Comment (sandeepsukhani) [reply pkg/loki/modules.go:549]: Prometheus: https://prometheus.io/docs/prometheus/latest/querying/api/#delete-series

Cortex: https://github.com/cortexproject/cortex/blob/6db67a4efbbf62b1133fa037a95382a21f752bbf/docs/guides/deleting-series.md

File: pkg/purger/http.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		if idx > 0 {
+			out.WriteRune(',')
+		}
+
+		out.WriteString(m.String())
+	}
+
+	out.WriteRune('}')
+	return out.String()
+}
```
Comment (cyriltovena) [inline pkg/purger/http.go:118]: nit: newline .

File: pkg/loki/modules.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/loki/modules.go:549]: Thanks I want to have a sense of the api request/responses.

File: pkg/purger/http.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	for i := range deleteRequests {
+		for j, selector := range deleteRequests[i].Selectors {
+
+			matchers, err := logql.ParseMatchers(selector)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (cyriltovena) [inline pkg/purger/http.go:83]: Can you log that as a warning ? this will help if it ever happens.

**Review by cyriltovena — COMMENTED** [review]
This is great work Sandeep.

I have 2 area of concern.

1. Cache

Do we also clear the cache (change the generation number) for a given user on delete cancellations ? I've tried to look at cortex code but I figured would be better to ask.

I don't see where the new header is set in the querier.

2. Iterators

I think pre- LogQL v2 this was fine to apply on the last iterators, although I think it would be slow as this is happening on the hot path (e.g for each entry).
Now the problem is that LogQL v2 has changed the game and labels can be mutated. This means deleted data could have its labels changed by a query and will pass your delete matcher filter although it is supposed to be deleted.
If you agree with me on this, I'd suggest we move this logic into the store and ingesters instead and cut the data before it gets processed. This means we will remove chunks (store) and streams (ingester), well done I'm sure this will also be faster.

WDYT ?

**Comment by sandeepsukhani** [comment]
> This is great work Sandeep.
> 
> I have 2 area of concern.
> 
> 1. Cache
> 
> Do we also clear the cache (change the generation number) for a given user on delete cancellations ? I've tried to look at cortex code but I figured would be better to ask.
> 
> I don't see where the new header is set in the querier. We use a different type of sharding middleware too, not sure it propagates this new header correctly.
Yes, cancellation changes the gen number. I just remembered I need to change the querier code to set an extra HTTP header for query-frontend, which I guess is missed. Will check and take care of it.
> 
> 1. Iterators
> 
> I think pre- LogQL v2 this was fine to apply on the last iterators, although I think it would be slow as this is happening on the hot path (e.g for each entry).
> Now the problem is that LogQL v2 has changed the game and labels can be mutated. This means deleted data could have its labels changed by a query and will pass your delete matcher filter although it is supposed to be deleted.
> If you agree with me on this, I'd suggest we move this logic into the store and ingesters instead and cut the data before it gets processed. This means we will remove chunks (store) and streams (ingester), well done I'm sure this will also be faster.
> 
> WDYT ?

I went with this approach because it was easiest and less complex but as you said if labels can change then there is a chance of incorrect results being returned. I will consider changing this to fix the problem that you mentioned.

Thanks for the review!

**Comment by cyriltovena** [comment]
> I went with this approach because it was easiest and less complex but as you said if labels can change then there is a chance of incorrect results being returned. I will consider changing this to fix the problem that you mentioned.


At some point we're going to have to change the interface of the iterator to add the base labels (from the chunk) which I think you could use. But I still think taking care of deletes at the chunks level is better. What's happening if you deletes data which is only half a chunk ? Do you rewrite the chunk ?

**Comment by sandeepsukhani** [comment]
> > I went with this approach because it was easiest and less complex but as you said if labels can change then there is a chance of incorrect results being returned. I will consider changing this to fix the problem that you mentioned.
> 
> At some point we're going to have to change the interface of the iterator to add the base labels (from the chunk) which I think you could use. But I still think taking care of deletes at the chunks level is better. What's happening if you deletes data which is only half a chunk ? Do you rewrite the chunk ?

Yes, we slice the chunk and upload it and delete the original chunk.
I think I missed adding support for that in Loki :facepalm:

**Comment by sandeepsukhani** [comment]
Closing this PR in favour of PR #3546, which adds support for deletion without query time filtering, which is complex to support in Loki at the moment.

**Pull request closed without merging**